### PR TITLE
Support batched input for decoder start ids

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

This PR addresses [issue #28763 ](https://github.com/huggingface/transformers/issues/28763). The requested feature already work out of the box, I just made it explicit and added one line in the docs. 

The changes were tested with `pytest -k generate_input tests/generation/test_utils.py`

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).

## Who can review?
@gante 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -515,6 +515,8 @@ def _prepare_decoder_input_ids_for_generation(
         decoder_start_token_id = self._get_decoder_start_token_id(decoder_start_token_id, bos_token_id)
         if device is None:
             device = self.device
+        if isinstance(decoder_start_token_id, torch.Tensor) and decoder_start_token_id.shape != (batch_size, 1):
```
Comment (zucchini-nlp) [inline src/transformers/generation/utils.py:518]: I cannot think of cases when start_ids have length > 1

**Review by gante — APPROVED** [review]
LGTM, only two minor nits. Thanks for fixing!

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -527,7 +529,13 @@ def _prepare_decoder_input_ids_for_generation(
             pass
         # user input but doesn't start with decoder_start_token_id -> prepend decoder_start_token_id (and adjust
         # decoder_attention_mask if provided)
-        elif (decoder_input_ids[:, 0] != decoder_start_token_id).all().item():
+        if (
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py]: accidentally changed that, now it's back

**Comment by zucchini-nlp** [comment]
@amyeroberts or @ArthurZucker Hi, PR is ready to review

File: src/transformers/generation/configuration_utils.py [hunk 3]
```diff
@@ -233,8 +233,11 @@ class GenerationConfig(PushToHubMixin):
         encoder_no_repeat_ngram_size (`int`, *optional*, defaults to 0):
             If set to int > 0, all ngrams of that size that occur in the `encoder_input_ids` cannot occur in the
             `decoder_input_ids`.
-        decoder_start_token_id (`int`, *optional*):
-            If an encoder-decoder model starts decoding with a different token than *bos*, the id of that token.
+        decoder_start_token_id (`Union[int, torch.Tensor]`, *optional*):
+            If an encoder-decoder model starts decoding with a different token than *bos*, the id of that token or a tensor with shape
+            `(batch_size, 1)`. Indicating a tensor enables different start ids for each element in the batch
+            (e.g. multilingual models with different target languages in one batch)
```
Comment (amyeroberts) [inline src/transformers/generation/configuration_utils.py:239]: What happens when you create a config with a tensor, save it and load it again? 

I don't think tensors are json serializable, so you'll need to conversion before saving and loading.

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

You'll need to add tests for the batched vs unbatched behaviour, as well as serializing the config.

File: src/transformers/generation/configuration_utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/configuration_utils.py:239]: Oh yeah, that's a good point @amyeroberts ! Perhaps it being a `list` is wiser

**Comment by gante** [comment]
@amyeroberts it should be ready to merge -- with a `list` as input (as opposed to a tensor), there are no serialization issues 🤗

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for iterating and adding this!

**Comment by amyeroberts** [comment]
@gante Yep! I can merge

**Comment by nlaimr** [comment]
Can this be done for [tf_utils.py ](https://github.com/huggingface/transformers/blob/v4.46.0/src/transformers/generation/tf_utils.py#L645) as well? 

Currently, it is passed to `_prepare_decoder_input_ids_for_generation ` from `generation_config` in `generate()`


`decoder_start_token_id=generation_config.decoder_start_token_id,`

**Pull request merged**